subsys/bluetooth/lib/ead.c

```diff
@@ -0,0 +1,220 @@
+
+static int bt_ead_generate_randomizer(uint8_t randomizer[BT_EAD_RANDOMIZER_SIZE])
+{
+	int err;
+
+	err = bt_rand(randomizer, BT_EAD_RANDOMIZER_SIZE);
+
+	if (err != 0) {
+		return -ECANCELED;
+	}
+
+	/* From Supplement to the Bluetooth Core Specification v11, Part A 1.23.3: The directionBit
+	 * of the CCM nonce shall be set to the most significant bit of the Randomizer field.
+	 */
+	randomizer[4] |= 1 << BT_EAD_RANDOMIZER_DIRECTION_BIT;
+
+	return 0;
+}
+
+static int bt_ead_generate_nonce(const uint8_t iv[BT_EAD_IV_SIZE],
+				 const uint8_t randomizer[BT_EAD_RANDOMIZER_SIZE], uint8_t *nonce)
+{
+	uint8_t new_randomizer[BT_EAD_RANDOMIZER_SIZE];
+
+	if (randomizer == NULL) {
+		int err;
+
+		err = bt_ead_generate_randomizer(new_randomizer);
+
+		if (err != 0) {
+			LOG_ERR("Failed to generate Randomizer");
+			return -ECANCELED;
+		}
+
+		randomizer = new_randomizer;
+	}
+
+	memcpy(&nonce[0], randomizer, BT_EAD_RANDOMIZER_SIZE);
+	memcpy(&nonce[BT_EAD_RANDOMIZER_SIZE], iv, BT_EAD_IV_SIZE);
+
+	return 0;
+}
+
+size_t bt_get_bt_data_size(const struct bt_data *ad_structs, size_t ad_structs_len)
+{
+	size_t ad_size = 0;
+
+	for (size_t i = 0; i < ad_structs_len; i++) {
+		ad_size += sizeof(ad_structs[i].data_len) + sizeof(ad_structs[i].type) +
+			   ad_structs[i].data_len;
+	}
+
+	return ad_size;
+}
+
+size_t bt_data_to_ad_structure(const struct bt_data *input, uint8_t *output_ad_structure)
+{
+	uint8_t ad_data_len = input->data_len;
+	uint8_t data_len = ad_data_len + 1;
+
+	output_ad_structure[0] = data_len;
+	output_ad_structure[1] = input->type;
+
+	memcpy(&output_ad_structure[2], input->data, ad_data_len);
+
+	return data_len + 1;
+}
+
+static int ead_encrypt(const uint8_t session_key[BT_EAD_KEY_SIZE], const uint8_t iv[BT_EAD_IV_SIZE],
+		       const uint8_t randomizer[BT_EAD_RANDOMIZER_SIZE], const uint8_t *payload,
+		       size_t payload_size, uint8_t *encrypted_payload)
+{
+	int err;
+	uint8_t nonce[BT_EAD_NONCE_SIZE];
+	size_t ead_size = BT_EAD_RANDOMIZER_SIZE + payload_size + BT_EAD_MIC_SIZE;
+
+	err = bt_ead_generate_nonce(iv, randomizer, nonce);
+
+	if (err != 0) {
+		return -ECANCELED;
+	}
+
+	memcpy(encrypted_payload, nonce, BT_EAD_RANDOMIZER_SIZE);
+
+	err = bt_ccm_encrypt(session_key, nonce, payload, payload_size, bt_ead_aad, BT_EAD_AAD_SIZE,
+			     &encrypted_payload[BT_EAD_RANDOMIZER_SIZE], BT_EAD_MIC_SIZE);
+
+	if (err != 0) {
+		LOG_ERR("Failed to encrypt the payload (bt_ccm_encrypt err %d)", err);
+		return -EIO;
+	}
+
+	LOG_HEXDUMP_DBG(encrypted_payload, ead_size, "Encrypted Data: ");
+
+	return 0;
+}
+
+int bt_ead_encrypt(const uint8_t session_key[BT_EAD_KEY_SIZE], const uint8_t iv[BT_EAD_IV_SIZE],
+		   const uint8_t *payload, size_t payload_size, uint8_t *encrypted_payload)
+{
+	if (session_key == NULL || iv == NULL || payload == NULL || encrypted_payload == NULL) {
+		LOG_ERR("At least one of the argument is a NULL pointer.");
+		return -EINVAL;
+	}
```

Comment:
I'd suggest something like this:
```suggestion
	CHECKIF(session_key == NULL) { 
		LOG_DBG("session_key is NULL");
		return -EINVAL;
	}

	CHECKIF(iv == NULL) {
		LOG_DBG("iv is NULL);
		return -EINVAL;
	}

	CHECKIF(payload == NULL) {
		LOG_DBG("payload is NULL);
		return -EINVAL;
	}

	CHECKIF(encrypted_payload == NULL) {
		LOG_DBG("encrypted_payload is NULL);
		return -EINVAL;
	}
```

Since we are returning an error, we shouldn't use LOG_ERR. The splitting of the validation checks make it much easier to, when LOG_DBG is enabled, to see exactly which one of them is failing. Finally, using CHECKIF allows the user to disable this check or make it an ASSERT if wanted.